[JSC] Use shuffle algorithm from BBQ for CCallHelpers::shuffleRegisters

ddegazio opened the pull request:
#### 40c330f75da3abf2cad6a49e51ec81e7adce87e9
<pre>
[JSC] Use shuffle algorithm from BBQ for CCallHelpers::shuffleRegisters
<a href="https://bugs.webkit.org/show_bug.cgi?id=289306">https://bugs.webkit.org/show_bug.cgi?id=289306</a>
<a href="https://rdar.apple.com/146442401">rdar://146442401</a>

Reviewed by Yusuke Suzuki.

Ports the shuffle algorithm used in the Wasm BBQ JIT to CCallHelpers::shuffleRegisters.
This results in fewer redundant moves being generated, particularly due to the fact we
are no longer using swaps to resolve dependent move cycles. Using this algorithm, we
elide 17234 out of 160180 moves total resolving shuffles on JetStream 2, a 10.7%
reduction, reducing code size and avoiding cheap but redundant work.

* Source/JavaScriptCore/jit/CCallHelpers.h:
(JSC::CCallHelpers::emitShuffleMove):
(JSC::CCallHelpers::scratchRegisterByType&lt;RegisterID&gt;):
(JSC::CCallHelpers::scratchRegisterByType&lt;FPRegisterID&gt;):
(JSC::CCallHelpers::shuffleRegisters):
* Source/JavaScriptCore/wasm/WasmBBQJIT.h:

Canonical link: <a href="https://commits.webkit.org/291786@main">https://commits.webkit.org/291786@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a41db803f50ff8f8f4c7b714542c07bbe248d186

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/93923 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/13508 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/3243 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/98929 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/44450 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/13805 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/21939 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/71673 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/29022 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/96925 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/10247 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/84848 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/52011 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/9929 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/2489 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/43766 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/86633 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/80192 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/2571 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/100969 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/92589 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/20975 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/15292 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/80690 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/21227 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/80813 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/80044 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19950 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/24583 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/1943 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/14132 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/20959 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/26137 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/115239 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/20646 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/33130 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/24106 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/22387 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/jit/CCallHelpers.h:
```diff
@@ -88,6 +89,50 @@ class CCallHelpers : public AssemblyHelpers {
+        status[index] = ShuffleStatus::BeingMoved;
+        for (unsigned i = 0; i < moves.size(); i ++) {
+            if (moves[i].first == moves[index].second) {
+                ASSERT(i != index);
```
Is it ensured? I wonder if we should have early check like what we have in BBQJIT::emitShuffleMove.

ddegazio replied:
It should be ensured, in the existing implementation of `shuffleRegisters` we have:
```
for (unsigned i = 0; i < NumberOfRegisters; ++i) {
    if (sources[i] != destinations[i])
        pairs.append(std::make_pair(sources[i], destinations[i]));
}
```
so trivial moves are already checked for and shouldn't make it into the shuffle algorithm.

ddegazio committed:
[JSC] Use shuffle algorithm from BBQ for CCallHelpers::shuffleRegisters
https://bugs.webkit.org/show_bug.cgi?id=289306
rdar://146442401

Reviewed by Yusuke Suzuki.

Ports the shuffle algorithm used in the Wasm BBQ JIT to CCallHelpers::shuffleRegisters.
This results in fewer redundant moves being generated, particularly due to the fact we
are no longer using swaps to resolve dependent move cycles. Using this algorithm, we
elide 17234 out of 160180 moves total resolving shuffles on JetStream 2, a 10.7%
reduction, reducing code size and avoiding cheap but redundant work.

* Source/JavaScriptCore/jit/CCallHelpers.h:
(JSC::CCallHelpers::emitShuffleMove):
(JSC::CCallHelpers::scratchRegisterByType<RegisterID>):
(JSC::CCallHelpers::scratchRegisterByType<FPRegisterID>):
(JSC::CCallHelpers::shuffleRegisters):
* Source/JavaScriptCore/wasm/WasmBBQJIT.h:

Canonical link: https://commits.webkit.org/291786@main
